core/vm/evm.go

```diff
@@ -426,6 +431,33 @@ func (evm *EVM) create(caller ContractRef, codeAndHash *codeAndHash, gas uint64,
 	if evm.StateDB.GetNonce(address) != 0 || (contractHash != (common.Hash{}) && contractHash != emptyCodeHash) {
 		return nil, common.Address{}, 0, ErrContractAddressCollision
 	}
+
+	// Initialise a new contract and set the code that is to be used by the EVM.
+	// The contract is a scoped environment for this execution context only.
+	contract := NewContract(caller, AccountRef(address), value, gas)
+	contract.SetCodeOptionalHash(&address, codeAndHash)
+
+	var (
+		isCallerEOF   = fromEOF
+		isInitcodeEOF = hasEOFMagic(codeAndHash.code)
+	)
+	if evm.chainRules.IsShanghai {
+		if isCallerEOF && !isInitcodeEOF {
+			// Don't allow EOF contract to run legacy initcode.
+			return nil, common.Address{}, gas, ErrLegacyCode
+		} else if isInitcodeEOF {
+			// If the initcode is EOF, verify it is well-formed.
+			var c Container
+			if err := c.UnmarshalBinary(codeAndHash.code); err != nil {
+				return nil, common.Address{}, gas, fmt.Errorf("%v: %v", ErrInvalidEOF, err)
+			}
+			if err := c.ValidateCode(evm.interpreter.cfg.JumpTableEOF); err != nil {
```

Comment:
EOF validation of initcode should be done before caller nonce is updated. In case it's invalid nonce should not be updated. Am I right?

Reply:
> Am I right?

Yes, I agree

Reply:
If so it should be also changed in tests. I.e. here https://github.com/ethereum/tests/blob/develop/src/GeneralStateTestsFiller/EIPTests/stEOF/stEIP3540/CREATE2_EOF1InvalidFiller.yml#L423

Reply:
Also agree. I updated here: e1e568b3b

One question - should nonce update also be omitted if EOF contract tries to run legacy initcode? I think yes, but it isn't clear.

Reply:
Answer is yes, nonce update should be omitted.